### ISSUE TITLE
')' character inserted into terminal after launching fish

### ISSUE THREAD

**Issue by nwoeanhinnogaehr** [fish-shell/fish-shell#3205]
- [X] Did you try [a quick search?](/issues?utf8=✓&q=is:issue+user:fish-shell)
- [X] Try [with the latest](/fish-shell/fish-shell/releases/tag/2.3.1) fish version?

When `__fish_vcs_prompt` is used in `fish_prompt` and `fish_greeting` is empty, a `)` character is inserted into the terminal after the prompt. It is actually part of the command line, and must be deleted before commands can be run.

**fish verison installed**: `fish, version 2.3.1`
**OS/terminal used**: Arch Linux, rxvt-unicode v9.22

## Reproduction steps

1. Use a prompt that includes `__fish_vcs_prompt` such as lonetwin, Terlar, Classic + Vcs, or Informative Vcs.
2. Put either `set fish_greeting ""` or just `set fish_greeting` in your config.fish.
3. Launch fish.

## Expected results
`user@host ~ (master)> `

## Actual results
`user@host ~ (master)> )`

**Comment by krader1961** [comment]
I can't reproduce and it seems odd that no other developers (who almost certainly all use `__fish_vcs_prompt` in their prompt) haven't noticed this problem. Are you using a plugin manager like oh-my-fish or fisherman? Do you have any custom scripts in *~/.config/fish/functions* that might be relevant? Can you reproduce if you replace your *~/.config/fish/config.fish* with one that contains nothing but `function fish_greeting; end`? If your *~/.config/fish/functions/fish_prompt.fish* script isn't an unadulterated copy of a stock prompt does the problem go away if you replace it with a stock prompt?

**Comment by faho** [comment]
@nwoeanhinnogaehr (I have to say I'm grateful for github's completion there):

What's the output if you execute `__fish_vcs_prompt` manually? Is this dependent on the directory you start fish in? If you `cd` somewhere else, does the ")" reappear? Which of the supported vcsen (svn, hg, git) do you have installed?

**Comment by nwoeanhinnogaehr** [comment]
I can't seem to reproduce it on my other machine either...

I'm not using any plugins. The only thing in the functions directory is the prompt script, which was chosen by fish_config. Replacing config.fish with just that one line has no effect. I also deleted my fishd.hostname file afterwards.

The character only appears when the shell is first launched. Once I delete it it's gone until I restart the shell. There is a git repo in my home directory, but removing it has no effect.

It appears to only be happening with urxvt. If I login to a virtual console or use xterm, the problem does not occur. It might be something strange with my urxvt config then.

**Comment by faho** [comment]
@nwoeanhinnogaehr: 

The most likely offender for me seems to be the svn-prompt, since it has the line `printf ')'`. Can you confirm this still occurs if you replace the `fish_vcs_prompt` calls with `__fish_svn_prompt`? Do you have svn installed? If so, what is the output of `svn status`?

**Comment by nwoeanhinnogaehr** [comment]
I've tracked it down to `__fish_hg_prompt`. Using either `__fish_svn_prompt` or `__fish_git_prompt` works fine. I have svn installed, but do not have mercurial installed.

**Comment by faho** [comment]
Can you confirm you are using a pristine copy of `__fish_hg_prompt`?

This is quite puzzling, as the first lines of that function are:

```fish
    if not command -s hg > /dev/null
        return 1
    end
```

And that obviously prints nothing, and it especially does _not_ enter anything into your commandline.

It sets some variables before that, but all of the values are simple and also seem "obviously" okay.

What _might_ be happening is that your terminal interprets something wrong, though I'm not even sure what that something is supposed to be.

**Comment by krader1961** [comment]
> What might be happening is that your terminal interprets something wrong, ...

That was my thought as well. There are escape sequences that interrogate the terminal and cause it to inject characters into the tty. For example, you can ask the terminal to report the cursor col and row. This page is the most comprehensive listing of such sequences I know of: http://ttssh2.osdn.jp/manual/en/about/ctrlseq.html.

**Comment by nwoeanhinnogaehr** [comment]
I reinstalled fish, just to be sure.

Some further investigation has revealed that at least one of the `fish_color_hg_*` variables must be set for the issue to occur. So, I checked my locale info and apparently nothing was set properly. `$LANG` was empty and everything else set to `POSIX`... Putting `LANG=en_US.UTF-8` in `/etc/environment` seems to have fixed it.

Still I'm not sure if that fully explains why this was happening. Surely those color variables would never even be read? And why just with the hg prompt?

**Comment by faho** [comment]
>Still I'm not sure if that fully explains why this was happening. Surely those color variables would never even be read? And why just with the hg prompt?

It actually _kinda_ does - at least I can sort of see a path for it to occur.

The function file is in UTF-8, which is a superset of ASCII - some characters (the multi-byte ones) have the high-bit set, which means they don't work as ASCII characters. There are some examples of these characters in those variables:

```fish
set -g fish_prompt_hg_status_added '✚'
set -g fish_prompt_hg_status_copied '⇒'
set -g fish_prompt_hg_status_deleted '✖'
```

Now it seems there's some case in fish where some buffer overflow or something occurs and something ends up in the commandline.

I'm not sure this is worth fixing - we can't figure out what encoding a file is, so we have to trust the system, and if you set your encoding to ASCII you get what you deserve.

>So, I checked my locale info and apparently nothing was set properly

Let me guess - you eschew a DM in favor of `startx` or similar, and you use systemd? See #277 (TL;DR: systemd unsets locale for getty, and other shells then re-set it via locale.sh - we don't, because we can't read that).

I'm growing more convinced that this is something that hits real users, and is a real annoyance.

Everyone okay with fixing just #277, which would fix 99% of this?

**Comment by krader1961** [comment]
I'm okay with "fixing" issue #277. Note the scare quotes. I'm worried that implicitly setting the locale vars while initializing fish won't introduce other problems. But fixing the problems we know about in this regard almost certainly justifies potentially introducing hypothetical problems. I would also like confirmation from @nwoeanhinnogaehr they are starting a graphical UI using `startx` or equivalent command from their text based console UI.

Having said that I'm not convinced fixing #277 is the correct solution to this issue. The "C" locale should correctly handle the multibyte UTF-8 chars in the *__fish_hg_prompt.fish* script. Setting an explicit UTF-8 ctype locale may solve this issue but that shouldn't be necessary. Such a solution may simply paper over a subtle, but significant, bug.

**Comment by nwoeanhinnogaehr** [comment]
That's exactly correct, I'm using systemd and startx.

**Comment by krader1961** [comment]
@nwoeanhinnogaehr I've committed a fix for issue #277. If you're able to build from git head and verify that fixes your problem that would be much appreciated.

**Comment by nwoeanhinnogaehr** [comment]
I can confirm it works. Thank you!
